autoscaler/collectormetrics/run.go

```diff
@@ -0,0 +1,181 @@
+)
+
+type MetricFetchResult struct {
+	PodName string
+	Error   error
+	Metrics map[string]*dto.MetricFamily
+}
+
+func (a *Autoscaler) Run(ctx context.Context) {
+	logger := log.FromContext(ctx)
+	logger = logger.WithName("autoscaler")
+
+	for {
+		select {
+		case notification := <-a.notifications:
+			logger.V(5).Info("Got ip change notification", "notification", notification)
+			a.updateIPsMap(notification)
+		case <-ctx.Done():
+			logger.V(0).Info("Shutting down autoscaler", "collectorsGroup", a.options.collectorsGroup)
+			a.ticker.Stop()
+			close(a.notifications)
+			return
+		case <-a.ticker.C:
+			results := a.getCollectorsMetrics(ctx)
+			decision := a.options.algorithm.Decide(ctx, results)
+			a.executeDecision(ctx, decision, len(results))
+		}
+	}
+}
+
+func (a *Autoscaler) adjustDecisionToBounds(decision AutoscalerDecision) AutoscalerDecision {
+	current := int(decision)
+	if current < a.options.minReplicas {
+		return AutoscalerDecision(a.options.minReplicas)
+	} else if current > a.options.maxReplicas {
+		return AutoscalerDecision(a.options.maxReplicas)
+	}
+
+	return decision
+}
+
+func (a *Autoscaler) executeDecision(ctx context.Context, decision AutoscalerDecision, currentReplicas int) bool {
+	logger := log.FromContext(ctx)
+	newDecision := a.adjustDecisionToBounds(decision)
+	if newDecision != decision {
+		logger.V(0).Info("Decision out of bounds, adjusting", "old", decision, "new", newDecision)
+	}
+
+	if newDecision == AutoscalerDecision(currentReplicas) {
+		logger.V(0).Info("No need to scale", "current", currentReplicas)
+		return false
+	}
+
+	obj := a.getTargetKubernetesObject()
+	if obj == nil {
+		logger.Error(errObjectNotFoundForCollectorsGroup, "No target object found", "group", a.options.collectorsGroup)
+		return false
+	}
+
+	scale := &autoscalingv1.Scale{
+		Spec: autoscalingv1.ScaleSpec{
+			Replicas: int32(decision),
+		},
+	}
+
+	err := a.kubeClient.SubResource("scale").Update(ctx, obj, client.WithSubResourceBody(scale))
+	if err != nil {
+		logger.Error(err, "Failed to scale object", "object", obj)
+		return false
+	}
+
+	return true
+}
+
+func (a *Autoscaler) getTargetKubernetesObject() client.Object {
+	if a.options.collectorsGroup == odigosv1.CollectorsGroupRoleClusterGateway {
+		return &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      gateway.KubeObjectName,
+				Namespace: env.GetCurrentNamespace(),
+			},
+		}
+	}
+
+	return nil
+}
+
+func (a *Autoscaler) updateIPsMap(notification Notification) {
+	if notification.Reason == NewIPDiscovered {
+		a.podIPs[notification.PodName] = notification.IP
+	} else if notification.Reason == IPRemoved {
+		delete(a.podIPs, notification.PodName)
+	}
+}
+
+func (a *Autoscaler) getCollectorsMetrics(ctx context.Context) []MetricFetchResult {
+	logger := log.FromContext(ctx)
+	results := make(chan MetricFetchResult, len(a.podIPs))
+	timeoutCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
```

Comment:
Maybe the timeout should be a function of the interval duration. Intuitively it should be a small fraction of the interval since we want our measurement to represent the interval and not take a significant time of  what it measures if that makes sense.